src/hooks/UniswapV2AdapterHook.sol

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.26;
+
+import {IUniswapV2Factory} from "briefcase/protocols/v2-core/interfaces/IUniswapV2Factory.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {IUniswapV2Pair} from "briefcase/protocols/v2-core/interfaces/IUniswapV2Pair.sol";
+import {BaseLiquidityAdapterHook} from "../base/hooks/BaseLiquidityAdapterHook.sol";
+import {SwapParams} from "@uniswap/v4-core/src/types/PoolOperation.sol";
+import {UniswapV2Library} from "../libraries/UniswapV2Library.sol";
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {Currency, CurrencyLibrary} from "@uniswap/v4-core/src/types/Currency.sol";
+
+/// @title UniswapV2AdapterHook
+/// @author Uniswap V4 Hooks
+/// @notice Adapter hook that routes V4 swaps through Uniswap V2 liquidity pools
+/// @dev Enables V4 pools to leverage existing V2 liquidity without migration
+contract UniswapV2AdapterHook is BaseLiquidityAdapterHook {
+    error InvalidFee();
+    error InvalidTickSpacing();
+
+    uint32 constant POOL_FEE = 3000;
+    int24 constant POOL_TICK_SPACING = 1;
+
+    /// @notice The Uniswap V2 factory contract for accessing V2 pairs
+    IUniswapV2Factory public immutable v2Factory;
+    bytes32 constant UNISWAP_V2_INIT_CODE_HASH = hex"96e8ac4277198ff8b6f785478aa9a39f403cb768dd02cbee326c3e7da348845f";
+
+    /// @notice Initializes the Uniswap V2 adapter hook
+    /// @param _manager The Uniswap V4 pool manager contract
+    /// @param _v2Factory The Uniswap V2 factory contract address
+    constructor(IPoolManager _manager, IUniswapV2Factory _v2Factory) BaseLiquidityAdapterHook(_manager) {
+        v2Factory = _v2Factory;
+    }
+
+    /// @inheritdoc BaseLiquidityAdapterHook
+    /// @dev Routes exact input swaps through the corresponding V2 pair
+    function _swapExactInput(PoolKey calldata poolKey, SwapParams calldata params)
+        internal
+        override
+        returns (uint256 amountOut)
+    {
+        (
+            Currency tokenIn,
+            Currency tokenOut,
+            address pair,
+            uint256 reserveIn,
+            uint256 reserveOut,
+            uint256 amountSpecified
+        ) = _parseSwap(poolKey, params);
+        amountOut = UniswapV2Library.getAmountOut(amountSpecified, reserveIn, reserveOut);
+        (uint256 amount0Out, uint256 amount1Out) = params.zeroForOne ? (uint256(0), amountOut) : (amountOut, uint256(0));
+
+        // Sync output token balance before V2 swap
+        poolManager.sync(tokenOut);
+        // Transfer input tokens from V4 pool manager directly to V2 pair
+        _take(tokenIn, pair, amountSpecified);
```

Comment:
This means that tokenIn must exist in pool manager which might not be the case _always_

unless we force that the router send in the user's tokens to v4 before the swap is called

Reply:
Yeah this is the case with the token wrapper hooks as well (WETH Hook, wsteth hook). I think Universal Router sends user's tokens in first now so typically these work!